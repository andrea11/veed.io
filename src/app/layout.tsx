import './globals.css'
import Banner from "@/components/Banner";
import Header from "@/components/Header";

export const metadata = {
  title: 'Github Repositories',
  description: 'A simple app to search for Github repos',
}

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html lang="en" className="h-full">
      <body className="h-full"><Banner/><Header/>{children}</body>
    </html>
  )
}
