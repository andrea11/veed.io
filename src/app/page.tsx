import Link from "next/link";

export default function Home() {
    return (
        <div className="flex flex-col gap-y-10 justify-center h-full max-h-96 mx-auto max-w-7xl px-6 lg:px-8">
            <section className="items-stretch">
                <h2 className="text-4xl font-bold tracking-tight sm:text-6xl">Welcome to my page</h2>
                <p className="mt-6 text-lg leading-8 text-gray-500">
                    This is a simple page that uses the GitHub API to display a list of repositories.
                    You can navigate to the <Link href="/repositories"
                                                  className="text-blue-900 font-medium underline">Repositories
                    page</Link> to see the list of repositories. You can save your favourite repositories by clicking on
                    the
                    star icon.
                    You can check your favourites in the <Link href="/favourites"
                                                               className="text-blue-900 font-medium underline">Favourites
                    page</Link>.
                </p>
            </section>
        </div>
    )
}