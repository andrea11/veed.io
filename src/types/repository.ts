import {Endpoints} from "@octokit/types";

export type Repository = Endpoints["GET /search/repositories"]["response"]["data"]["items"][0]
export type Repositories = Endpoints["GET /search/repositories"]["response"]["data"]["items"]
export type RepositoriesResponseData = Endpoints["GET /search/repositories"]["response"]["data"]