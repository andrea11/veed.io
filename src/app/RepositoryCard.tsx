import {Endpoints} from "@octokit/types"
import { identicon } from 'minidenticons'

import Link from "next/link";
import Identicon from "@/app/Identicon";

type RepositoryCardProps = {
    repository: Endpoints["GET /search/repositories"]["response"]["data"]["items"][0]
}

function generateImage(name: string) {
    return identicon("")
}
export default function RepositoryCard({repository}: RepositoryCardProps) {
    return (
        <div key={repository.id} className="group relative">
            <div
                className="aspect-w-1 aspect-h-1 w-full overflow-hidden rounded-md bg-gray-200 group-hover:opacity-75 lg:aspect-none lg:h-72">
                <Identicon username={repository.name} />
            </div>
            <div className="mt-4 flex justify-between">
                <div>
                    <h3 className="text-sm text-gray-700">
                        <Link href={repository.html_url}>
                            <span aria-hidden="true" className="absolute inset-0"/>
                            {repository.full_name}
                        </Link>
                    </h3>
                    <p className="mt-1 text-sm text-gray-500">{repository.description}</p>
                </div>
                <p className="text-sm font-medium text-gray-900">{repository.stargazers_count}&#9733;</p>
            </div>
        </div>
    )
}