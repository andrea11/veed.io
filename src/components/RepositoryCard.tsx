"use client"
import Link from "next/link";
import Identicon from "@/components/Identicon";
import {RepositoryStarButton} from "@/components/RepositoryStarButton";
import {Repository} from "@/app/types";
import {useEffect, useState} from "react";

type RepositoryCardProps = {
    repository: Repository
}

const formatter = Intl.NumberFormat('en', {notation: 'compact'})

function getFavouriteRepositories(): Record<string, Repository> {
    return JSON.parse(localStorage.getItem("favourites") ?? "{}")
}

function saveFavouriteRepositories(favourites: Record<string, Repository>): void {
    localStorage.setItem("favourites", JSON.stringify(favourites))
}

function isStarred(repository: Repository): boolean {
    return getFavouriteRepositories()[repository.id] !== undefined
}

export default function RepositoryCard({repository}: RepositoryCardProps) {
    const [starred, setStarred] = useState(false)

    useEffect(() => {
        setStarred(isStarred(repository))
    }, [repository])
    const handleStarButtonClick = () => {
        const favourites = getFavouriteRepositories()
        if (starred) {
            delete favourites[repository.id]
        } else {
            favourites[repository.id] = repository
        }

        saveFavouriteRepositories(favourites)
        setStarred(!starred)
    };

    return (
        <div key={repository.id} className="group relative">
            <div
                className="relative aspect-w-1 aspect-h-1 w-full overflow-hidden rounded-md bg-gray-200 lg:aspect-none lg:h-72">
                <Identicon username={repository.name}/>
                <div className="absolute z-10 bottom-0 right-0 mr-2 mb-2">
                    <RepositoryStarButton onClick={handleStarButtonClick} enabled={starred}
                                          numberOfLikes={formatter.format(repository.stargazers_count)}/>
                </div>
            </div>
            <Link href={repository.html_url}>
                <div className="mt-4 flex justify-between">
                    <div>
                        <h3 className="text-sm text-gray-700">
                            <span aria-hidden="true" className="absolute"/>
                            {repository.full_name}
                        </h3>
                        <p className="mt-1 text-sm text-gray-500">{repository.description}</p>
                    </div>
                </div>
            </Link>
        </div>
    )
}