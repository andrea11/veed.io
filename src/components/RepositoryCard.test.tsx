import RepositoryCard from "@/components/RepositoryCard";
import {act, cleanup, render, waitFor} from "@testing-library/react";
import {Repository} from "@/app/types";

jest.mock("minidenticons", () => ({
    identicon: jest.fn()
}))

describe("RepositoryCard", () => {
    const spyOnLocalStorageGetItem = jest.fn()
    const spyOnLocalStorageSetItem = jest.fn()
    let repository: Repository

    beforeEach(() => {
        jest.resetAllMocks()
        const localStorage = {
            getItem: spyOnLocalStorageGetItem,
            setItem: spyOnLocalStorageSetItem,
            removeItem: jest.fn(),
            clear: jest.fn(),
            key: jest.fn(),
            length: 0
        }
        Object.defineProperty(window, "localStorage", {value: localStorage});

        repository = {
            id: 1252125,
            name: "name",
            full_name: "full_name",
            description: "description",
            html_url: "html_url",
            stargazers_count: 0,
            url: "url",
        } as Repository
    })

    afterEach(() => {
        cleanup()
    })

    it("renders correctly", () => {
        const {container} = render(<RepositoryCard repository={repository} key={repository.id}/>)
        expect(container).toMatchSnapshot()
    })

    it("should retrieve favourites from local storage and set the button to enable if it matches", async () => {
        spyOnLocalStorageGetItem.mockReturnValue(JSON.stringify({[repository.id]: repository}))
        const {getByTestId} = render(<RepositoryCard repository={repository} key={repository.id}/>)
        await waitFor(() => expect(getByTestId("repository-star")).toHaveClass("text-yellow-400"))
    })

    it("should save favourites to local storage on click", async () => {
        const {getByTestId} = render(<RepositoryCard repository={repository} key={repository.id}/>)
        act(() => getByTestId("repository-star-button").click())
        await waitFor(() => expect(spyOnLocalStorageSetItem).toHaveBeenCalled())
        expect(spyOnLocalStorageSetItem).toHaveBeenCalledWith("favourites", JSON.stringify({[repository.id]: repository}))
    })

    it("should toggle the button on multiple clicks", async () => {
        const {getByTestId} = render(<RepositoryCard repository={repository} key={repository.id}/>)
        for (let i = 1; i <= 3; i++) {
            act(() => getByTestId("repository-star-button").click())
            await waitFor(() => {
                expect(getByTestId("repository-star")).toHaveClass(i % 2 === 1 ? "text-yellow-400" : "text-grey-400")
            })
        }
    })
})