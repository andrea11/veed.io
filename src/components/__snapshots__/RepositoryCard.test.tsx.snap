// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryCard renders correctly 1`] = `
<div>
  <div
    class="group relative"
  >
    <div
      class="relative aspect-w-1 aspect-h-1 w-full overflow-hidden rounded-md bg-gray-200 lg:aspect-none lg:h-72"
    >
      <div />
      <div
        class="absolute z-10 bottom-0 right-0 mr-2 mb-2"
      >
        <button
          class="flex h-min w-min space-x-1 items-center rounded-full text-gray-400 hover:text-rose-600 bg-gray-700 hover:bg-rose-50 py-1 px-2 text-xs font-medium"
          data-testid="repository-star-button"
        >
          <svg
            aria-hidden="true"
            class="w-5 h-5 hover:text-current text-grey-400"
            data-testid="repository-star"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              First star
            </title>
            <path
              d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
            />
          </svg>
          <p
            class="font-semibold text-xs hover:text-current"
          >
            0
          </p>
        </button>
      </div>
    </div>
    <a
      href="html_url"
    >
      <div
        class="mt-4 flex justify-between"
      >
        <div>
          <h3
            class="text-sm text-gray-700"
          >
            <span
              aria-hidden="true"
              class="absolute"
            />
            full_name
          </h3>
          <p
            class="mt-1 text-sm text-gray-500"
          >
            description
          </p>
        </div>
      </div>
    </a>
  </div>
</div>
`;
